File: Leetcode/others/20.有效的括号.cpp
/**
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串，判断字符串是否有效。

有效字符串需满足：
左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
注意空字符串可被认为是有效字符串。
*/
class Solution {
public:
    bool isValid(string s) {
        stack<char> backets;
        for(int i=0;i<s.size();i++)
        {
            if(s[i]=='('||s[i]=='['||s[i]=='{')
            {
                backets.push(s[i]);
            }
            else
            {
                if(backets.size()!=0)//栈不为空
                {
                    if(s[i]==')'&&backets.top()!='(')   return false;
                    if(s[i]=='}'&&backets.top()!='{')   return false;
                    if(s[i]==']'&&backets.top()!='[')   return false;
                    backets.pop();
                }
                else return false;
            }
        }
        if(backets.size()!=0)   return false;
        return true;
    }
};

File: Leetcode/simple/20. 有效的括号.py
"""
给定一个只包括 '('，')'，'{'，'}'，'['，']' 的字符串 s ，判断字符串是否有效。

有效字符串需满足：

左括号必须用相同类型的右括号闭合。
左括号必须以正确的顺序闭合。
"""


class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if len(s)%2!=0:
            return False
        stack=[]
        for ch in s:
            if ch in ["(","{","["]:
                stack.append(ch)
            else:
                if len(stack)==0:
                    return False
                if ch==")" and stack[-1]=="(":
                    stack.pop()
                elif ch=="]" and stack[-1]=="[":
                    stack.pop()
                elif ch=="}" and stack[-1]=="{":
                    stack.pop()
                else:
                    return False
        if len(stack)!=0:
            return False
        else:
            return True